I couldn't find any relevant unittests (in fact, `qt/tests/` seems to only have a basic lint test for the whole `qt/` side; otherwise I'd have tried to at least add a unittest for `app.MainWindow.tmpCopy()` itself).

Did some manual testing.

Snapshot with regular symlink:
1. Setup:
   * Have backup source with at least one (valid) symlink
     ```sh
     $  ls -l backup_source/
     total 4
     lrwxrwxrwx 1 user users 3 11-19 16:58 bar -> foo
     -rw-r--r-- 1 user users 4 11-19 17:00 foo
     ```

   * Create 2 snapshots.  These do have the symlink in tact:
     ```sh
     $  ls -l backup_target/backintime/host/user/1/last_snapshot/backup/home/user/dev/external/backintime/test.no-git/fake_home/backup_source/
     backup_source/
     total 4
     lrwxrwxrwx 2 user users 3 11-19 16:58 bar -> foo
     -rw-r--r-- 1 user users 4 11-19 17:00 foo
     ```

2. Test without this fix
   * Compare the 2 snapshots in the GUI, using `Backup -> Compare snapshots...` (pick one snapshot from the list in the main view, pick the other from the drop-down right to the `Compare` button at the bottom)
   * Check temp copies in shell:
     ```sh
     $  ls -l /tmp/tmp*/backup_source
     /tmp/tmpedqq05c8_20241119-170042-790/backup_source:
     total 8
     -rw-r--r-- 1 user users 4 11-19 17:00 bar
     -rw-r--r-- 1 user users 4 11-19 17:00 foo

     /tmp/tmpjl7xix7a_20241119-165907-278/backup_source:
     total 0
     -rw-r--r-- 1 user users 0 11-19 16:58 bar
     -rw-r--r-- 1 user users 0 11-19 16:58 foo
     ```
   * The symlinks are now replaced with their targets (this results in a full disk if you have (eventually) circular symlinks, like something in your home pointing to `/` or even just your home)

3. Test with this fix
   * Compare the 2 snapshots, same way as above
   * Check in shell:
     ```sh
     $  ls -l /tmp/tmp*/backup_source/
     /tmp/tmp6f5bpc2q_20241119-170042-790/backup_source/:
     total 4
     lrwxrwxrwx 1 user users 3 11-19 16:58 bar -> foo
     -rw-r--r-- 1 user users 4 11-19 17:00 foo

     /tmp/tmpykrspljt_20241119-165907-278/backup_source/:
     total 0
     lrwxrwxrwx 1 user users 3 11-19 16:58 bar -> foo
     -rw-r--r-- 1 user users 0 11-19 16:58 foo
     ```

Snapshot with broken symlink (pointing to nonexistent target):
1. Setup
   * Have a backup source with a 'broken' symlink (pointing to a nonexistent target):
     ```sh
     $  ls -l backup_source/
     total 4
     lrwxrwxrwx 1 user users  3 11-19 16:58 bar -> foo
     lrwxrwxrwx 1 user users 11 11-21 11:26 baz -> nonexistent
     -rw-r--r-- 1 user users  4 11-19 17:00 foo
     ```

   * Create 2 snapshots.  These do have the symlink in tact:
     ```sh
     $  ls -l backup_target/backintime/host/user/1/last_snapshot/backup/home/user/dev/external/backintime/test.no-git/fake_home/backup_source/
     total 4
     lrwxrwxrwx 1 ephedio users  3 11-19 16:58 bar -> foo
     lrwxrwxrwx 1 ephedio users 11 11-21 11:26 baz -> nonexistent
     -rw-r--r-- 1 ephedio users  9 11-21 11:26 foo
     ```

2. Test without this fix
   * Compare the 2 snapshots, same way as above
   * App now crashes:
     ```
     Traceback (most recent call last):
       File "/home/user/dev/external/backintime/qt/snapshotsdialog.py", line 370, in btnDiffClicked
         path1 = self.parent.tmpCopy(path1, sid1)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       File "/home/user/dev/external/backintime/qt/app.py", line 1725, in tmpCopy
         shutil.copytree(full_path, tmp_file)
       File "/usr/lib64/python3.11/shutil.py", line 573, in copytree
         return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       File "/usr/lib64/python3.11/shutil.py", line 527, in _copytree
         raise Error(errors)
     shutil.Error: [('/home/user/dev/external/backintime/test.no-git/fake_home/backup_target/backintime/host/user/1/20241121-112648-601/backup/home/user/dev/external/backintime/test.no-git/fake_home/backup_source/baz', '/tmp/tmpvvv2e1nz_20241121-112648-601/backup_source/baz', "[Errno 2] No such file or directory: '/home/user/dev/external/backintime/test.no-git/fake_home/backup_target/backintime/host/user/1/20241121-112648-601/backup/home/user/dev/external/backintime/test.no-git/fake_home/backup_source/baz'")]
     /home/user/dev/external/backintime/test.no-git/../qt/backintime-qt: line 31:  5568 Aborted                 /usr/bin/env python3 -Es ${APP_PATH}/app.py "$@"
     ```

3. Test with this fix
   * Compare the 2 snapshots, same way as above
   * App now doesn't crash and has the broken symlink in tact as well:
     ```sh
     $  ls -l /tmp/tmp*/backup_source/
     /tmp/tmpcahq_g0n_20241121-112635-905/backup_source/:
     total 4
     lrwxrwxrwx 1 user users  3 11-19 16:58 bar -> foo
     lrwxrwxrwx 1 user users 11 11-21 11:26 baz -> nonexistent
     -rw-r--r-- 1 user users  4 11-19 17:00 foo

     /tmp/tmprs7_uauk_20241121-112648-601/backup_source/:
     total 4
     lrwxrwxrwx 1 user users  3 11-19 16:58 bar -> foo
     lrwxrwxrwx 1 user users 11 11-21 11:26 baz -> nonexistent
     -rw-r--r-- 1 user users  9 11-21 11:26 foo
     ```

--- a/CHANGES
+++ b/CHANGES
@@ -1,6 +1,8 @@
 Back In Time
 
 Version 1.6.0-dev (development of upcoming release)
+* Fix: Copy symlink as symlink in snapshot compare (#1902) (Peter Sevens @sevens)
+       Additionally fixes a crash when comparing a snapshot with a symlink pointing to a nonexistent target.
 * ...
 
 Version 1.5.3 (2024-11-13)

--- a/CHANGES
+++ b/CHANGES
@@ -1,8 +1,8 @@
 Back In Time
 
 Version 1.6.0-dev (development of upcoming release)
-* Fix: Copy symlink as symlink in snapshot compare (#1902) (Peter Sevens @sevens)
-       Additionally fixes a crash when comparing a snapshot with a symlink pointing to a nonexistent target.
+* Fix: Snapshot compare copy symlink as symlink (#1902) (Peter Sevens @sevens)
+* Fix: Crash when comparing a snapshot with a symlink pointing to a nonexistent target (Peter Sevens @sevens)
 * Fix: Crash (KeyError) opening language setup dialog with unknown locale/language
 
 Version 1.5.3 (2024-11-13)

--- a/qt/app.py
+++ b/qt/app.py
@@ -1722,7 +1722,7 @@ class MainWindow(QMainWindow):
         tmp_file = os.path.join(d.name, os.path.basename(full_path))
 
         if os.path.isdir(full_path):
-            shutil.copytree(full_path, tmp_file)
+            shutil.copytree(full_path, tmp_file, symlinks=True)
         else:
             shutil.copy(full_path, d.name)
 

